### PR TITLE
feat: add npm package overrides to frontend builder

### PR DESCRIPTION
**Pull request by abutterworth**
Consume config to override npm packages, leveraging npm aliases.

Sample config in edx-internal: https://github.com/edx/edx-internal/blob/pipeline-test/frontends/frontend-app-profile/stage_config.yml
```
APP_CONFIG:
  # 
  # 
NPM_ALIASES:
  "@edx/frontend-component-open-edx-header": "npm:@edx/frontend-component-edx-header"
```

### CODE REVIEW

File: tubular/scripts/frontend_utils.py [hunk 1]
```diff
@@ -53,10 +53,19 @@ def _get_configs(self):
 
     def install_requirements(self):
         """ Install requirements for app to build """
-        proc = subprocess.Popen(['npm install'], cwd=self.app_name, shell=True)
+        proc = subprocess.Popen(['npm install -g npm && npm install'], cwd=self.app_name, shell=True)
```
Comment (douglashall) [inline tubular/scripts/frontend_utils.py]: What does this do?

File: tubular/scripts/frontend_utils.py [hunk 2]
```diff
@@ -53,10 +53,19 @@ def _get_configs(self):
...
+            self.FAIL('Could not run `npm install -g npm && npm install` for app {}.'.format(self.app_name))
+        npm_overrides = self.get_override_config()
+        if npm_overrides:
+            aliased_installs = ' '.join(['{}@{}'.format(k, v) for k, v in npm_overrides.items()])
+            override_proc = subprocess.Popen(['npm install {}'.format(aliased_installs)], cwd=self.app_name, shell=True)
+            override_proc_return_code = override_proc.wait()
+            if override_proc_return_code != 0:
+                self.FAIL('Could not run `npm install` overrides {} for app {}.'.format(
+                    aliased_installs, self.app_name
+                ))
```
Comment (douglashall) [inline tubular/scripts/frontend_utils.py:84]: Maybe move this to a separate function `install_requirements_overrides` with a docstring explaining what's going on here.

File: tubular/scripts/frontend_utils.py [hunk 3]
```diff
@@ -66,6 +75,14 @@ def get_app_config(self):
             self.LOG('Config variables do not exist for app {}.'.format(self.app_name))
         return app_config
 
+    def get_override_config(self):
```
Comment (douglashall) [inline tubular/scripts/frontend_utils.py]: `get_requirements_override_config`

**Review by douglashall — APPROVED** [review]

File: tubular/scripts/frontend_utils.py (hunk 2 above)
Comment (tuchfarber) [reply tubular/scripts/frontend_utils.py:84]: Agreed on the above. Also, not entirely sure the process of installing these aliases. Are we expecting a base version to be installed, then these install over it? I'm trying to think through if there's going to be any issues (performance, side-effect, etc) of running the two version of `npm install`

**Review by tuchfarber — COMMENTED** [review]
Mostly just agreeing with Doug's points, as far as structure.

I don't particularly like the work `override` being used as the whole system (esp. when we get into multi-deployment) is a series of layered overrides. Under the assumption that this new key and code is going to be used just for npm aliases, could we more specifically call it NPM_ALIASES and use that terminology throughout the code? Less time groking what's an alias override vs an app_config override.

File: tubular/scripts/frontend_utils.py (hunk 1 above)
Comment (abutterworth) [reply tubular/scripts/frontend_utils.py]: We're having trouble upgrading node and npm in the go-agent Dockerfiles. This is a temporary work around to upgrade npm without needing the get that to work first. Less than ideal, but the easiest path forward for now.

File: tubular/scripts/frontend_utils.py (hunk 2 above)
Comment (abutterworth) [reply tubular/scripts/frontend_utils.py:84]: We're expecting a default, open-edx version of packages to be installed. To your concerns:
 - Performance-wise the only issue here is that we'll slow down frontend builds
 - Side-effects: I think npm should re-evaluate the dependency tree given a new package to consume. This should be okay assuming we are thoughtful about using peer-dependencies in these swappable packages. If we aren't careful it is possible we could get duplicate versions of react in our builds.

File: tubular/scripts/frontend_utils.py (hunk 1 above)
Comment (douglashall) [reply tubular/scripts/frontend_utils.py]: I'd probably wait on the node upgrade, unless there is a pressing need to get this change in before that.

File: tubular/scripts/frontend_utils.py (hunk 2 above)
Comment (davidjoy) [reply tubular/scripts/frontend_utils.py:84]: One thing that might make the code a bit more consistent / remove some ambiguity around npm:

1. Upgrade npm using `npm install -g npm` prior to using it for the first time in `install_requirements`.
2. Then you can just call the global `npm install` in both `install_requirements` and in `install_requirements_npm_aliases`.

That way you're not using two separate versions of `npm install`, one from global, and one from node_modules/.bin`.  I'd be worried that if they're different enough we might see some strange interactions, as NPM is pretty notorious for changing things version to version.

To @tuchfarber's comment about performance w/r/t using `npm install` multiple times - the first call will install everything in package.json, and the call in `install_requirements_npm_aliases` is doing targeted installs of specific packages.  So it shouldn't be doing unnecessary work or doing the same work twice.

Comment (abutterworth) [reply tubular/scripts/frontend_utils.py:84]: > Upgrade npm using npm install -g npm prior to using it for the first time in install_requirements.

We can't install anything globally from this context (for lack of permissions).

**Pull request merged**